### DISCUSSION TITLE
How-to use caching

### DISCUSSION THREAD

**Discussion by glenthomas** [dflook/terraform-github-actions#262] [General]
I am currently looking at ways to reduce GitHub Actions workflow durations and one possible optimisation could be to cache the Terraform CLI, provider plugins and modules.

My understanding is that:
- the Terraform CLI is downloaded from Hashicorp and installed on every run.
- the provider plugins are downloaded from the provider registries and installed on every run.
- the modules are downloaded from their sources on every run.

I am expecting that their may be some performance benefit (maybe a small one) in saving these to the GitHub Actions cache between runs to avoid the multiple downloads that take place.

Workflow caching for Terraform is not something that I have attempted before...

Does anyone have experience of implementing any of this with the dflook actions?
Does anyone have any opinions on why it might not be a good idea or not be possible?

**Comment by dflook** [comment]
The actions try and cache the terraform binary and plugins on disk, but if you are using github hosted runners (or ephemeral self-hosted runners) the cache will start empty for each job.

There was previous discussion in #105. Essentially, If you know the path to the cache on disk you can use it with [actions/cache](https://github.com/actions/cache) to save and restore those paths and carry the cache between jobs.

Today, On github hosted runners you should be able to cache terraform plugins using the path:
`${{ runner.temp }}/_github_home/.terraform.d/plugin-cache`
and terraform binaries using
`${{ runner.temp }}/_github_home/.dflook-terraform-github-actions`

This shouldn't be two paths - in the future everything will be moved under the second path. I haven't actually tested this myself so it would be great to see a working example.

**Reply by spencerfong-iceye** [reply]
I am coming across this same use-case a few years later, but I am unable to get caching working properly. Specifically, the cache does get uploaded, but when restoring the cache it is ignored by Terraform and all plugins/providers are redownloaded from scratch.

Has there been any update with regards to caching functionality with these Actions? FWIW I am copying the supposedly working example that was posted in #105.
